src/pydvl/value/sampler.py

```diff
@@ -0,0 +1,290 @@
+        self._n = len(indices)
+        self._n_samples = 0
+
+    @property
+    def indices(self) -> NDArray[T]:
+        return self._indices
+
+    @indices.setter
+    def indices(self, indices: NDArray[T]):
+        raise AttributeError("Cannot set indices of sampler")
+
+    @property
+    def n_samples(self) -> int:
+        return self._n_samples
+
+    @n_samples.setter
+    def n_samples(self, n: int):
+        raise AttributeError("Cannot reset a sampler's number of samples")
+
+    def complement(self, exclude: Sequence[T]) -> NDArray[T]:
+        return np.setxor1d(self._indices, exclude)
+
+    def iterindices(self) -> Iterator[T]:
+        """Iterates over indices in the order specified at construction.
+
+        FIXME: this is probably not very useful, but I couldn't decide
+          which method is better
+        """
+        if self._index_iteration is PowersetSampler.IndexIteration.Sequential:
+            for idx in self._outer_indices:
+                yield idx
+        elif self._index_iteration is PowersetSampler.IndexIteration.Random:
+            while True:
+                yield np.random.choice(self._outer_indices, size=1).item()
+
+    @overload
+    def __getitem__(self, key: slice) -> "PowersetSampler[T]":
+        ...
+
+    @overload
+    def __getitem__(self, key: list[int]) -> "PowersetSampler[T]":
+        ...
+
+    def __getitem__(self, key: slice | list[int]) -> "PowersetSampler[T]":
+        if isinstance(key, slice) or isinstance(key, Iterable):
+            return self.__class__(
+                self._indices,
+                index_iteration=self._index_iteration,
+                outer_indices=self._indices[key],
+            )
+        raise TypeError("Indices must be an iterable or a slice")
+
+    def __len__(self) -> int:
+        """Returns the number of outer indices over which the sampler iterates."""
+        return len(self._outer_indices)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}"
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._indices})"
+
+    @abc.abstractmethod
+    def __iter__(self) -> Iterator[SampleType]:
+        ...
+
+    @abc.abstractmethod
+    def weight(self, subset: NDArray[T]) -> float:
+        r"""Factor by which to multiply Monte Carlo samples, so that the
+        mean converges to the desired expression.
+
+        By the Law of Large Numbers, the sample mean of $\delta_i(S_j)$
+        converges
+        to the expectation under the distribution from which $S_j$ is sampled.
+
+        $$ \frac{1}{m}  \sum_{j = 1}^m \delta_i (S_j) c (S_j) \longrightarrow
+           \underset{S \sim \mathcal{D}_{- i}}{\mathbb{E}} [\delta_i (S) c (
+           S)]$$
+
+        We add a factor $c(S_j)$ in order to have this expectation coincide with
+        the desired expression.
+        """
+        ...
+
+
+class DeterministicSampler(PowersetSampler[T]):
```

Comment:
Perhaps we should alias this as `CombinatorialSampler`?

Reply:
For the sake of completeness, shouldn't we also create a deterministic sampler for permutation sampling, too?

Reply:
Done in 2685c969c14b9a73847a317754a260910d76d267